File: src/components/pages/projects/$projectId/shared/connect-snippets/android/java/SignIn.java
import android.content.Context;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import io.appwrite.coroutines.CoroutineCallback;
import io.appwrite.services.Account;

public final class SignIn {
  private SignIn() {}

  public static View create(
      Context context, Runnable onSignedIn, Runnable onGoToSignUp) {
    LinearLayout layout = new LinearLayout(context);
    layout.setOrientation(LinearLayout.VERTICAL);
    int pad = (int) (24 * context.getResources().getDisplayMetrics().density);
    layout.setPadding(pad, pad, pad, pad);

    TextView heading = new TextView(context);
    heading.setText("Sign in");

    TextView error = new TextView(context);
    error.setVisibility(View.GONE);

    EditText email = new EditText(context);
    email.setHint("Email");
    email.setInputType(InputType.TYPE_CLASS_TEXT
        | InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);

    EditText password = new EditText(context);
    password.setHint("Password");
    password.setInputType(InputType.TYPE_CLASS_TEXT
        | InputType.TYPE_TEXT_VARIATION_PASSWORD);

    Button submit = new Button(context);
    submit.setText("Sign in");
    submit.setAllCaps(false);
    submit.setOnClickListener(v -> {
      String emailValue = email.getText().toString();
      String passwordValue = password.getText().toString();
      if (emailValue.isEmpty() || passwordValue.isEmpty()) return;
      error.setVisibility(View.GONE);

      Account account = new Account(AppwriteClient.get(context));
      try {
        account.createEmailPasswordSession(emailValue, passwordValue,
            new CoroutineCallback<>((session, e) -> layout.post(() -> {
              if (e == null) {
                onSignedIn.run();
              } else {
                showError(error, e, "Sign in failed");
              }
            })));
      } catch (Exception ignored) {}
    });

    TextView hint = new TextView(context);
    hint.setText("No account?");

    Button goToSignUp = new Button(context);
    goToSignUp.setText("Sign up");
    goToSignUp.setAllCaps(false);
    goToSignUp.setOnClickListener(v -> onGoToSignUp.run());

    layout.addView(heading);
    layout.addView(error);
    layout.addView(email);
    layout.addView(password);
    layout.addView(submit);
    layout.addView(hint);
    layout.addView(goToSignUp);
    return layout;
  }

  private static void showError(TextView error, Throwable e, String fallback) {
    error.setText(e.getMessage() != null ? e.getMessage() : fallback);
    error.setVisibility(View.VISIBLE);
  }
}

File: src/components/pages/projects/$projectId/shared/connect-snippets/android/kotlin/SignIn.kt
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.foundation.text.KeyboardOptions
import io.appwrite.services.Account
import kotlinx.coroutines.launch

@Composable
fun SignIn(onSignedIn: () -> Unit, onGoToSignUp: () -> Unit) {
  val context = LocalContext.current
  val scope = rememberCoroutineScope()
  var email by remember { mutableStateOf("") }
  var password by remember { mutableStateOf("") }
  var error by remember { mutableStateOf("") }

  Text("Sign in")
  if (error.isNotEmpty()) Text(error)

  TextField(
    value = email,
    onValueChange = { email = it },
    placeholder = { Text("Email") },
    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
  )
  TextField(
    value = password,
    onValueChange = { password = it },
    placeholder = { Text("Password") },
    visualTransformation = PasswordVisualTransformation(),
  )

  TextButton(
    onClick = {
      if (email.isEmpty() || password.isEmpty()) return@TextButton
      error = ""
      scope.launch {
        try {
          Account(AppwriteClient.get(context))
            .createEmailPasswordSession(email, password)
          onSignedIn()
        } catch (e: Exception) {
          error = e.message ?: "Sign in failed"
        }
      }
    },
  ) { Text("Sign in") }

  Text("No account?")
  TextButton(onClick = onGoToSignUp) { Text("Sign up") }
}
